independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,122 @@
+     *         accessible.
+     */
+    private JarFile acquire() {
+        while (true) {
+            int count = referenceCounter.get();
+            if (count == 0) {
+                return null;
+            }
+            if (referenceCounter.compareAndSet(count, count+1)) {
+                return jarFile;
+            }
+        }
+    }
+
+    /**
+     * Decrease the readers counter of the jarFile.
+     * If the counter drops to 0 and a release has been requested also closes the jarFile.
+     */
+    private void release() {
+        while (true) {
+            int count = referenceCounter.get();
+            if (count <= 0) {
+                throw new IllegalStateException("The reference counter cannot be negative, found: " + (referenceCounter.get() - 1));
+            }
+            if (referenceCounter.compareAndSet(count, count-1)) {
+                if (referenceCounter.get() == 0) {
+                    try {
+                        jarFile.close();
+                        jarFileReference.compareAndSet(this, null);
+                    } catch (Throwable e) {
+                        //ignore
+                    }
+                }
+                return;
+            }
+        }
+    }
+
+    /**
+     * Ask to close this reference.
+     * If there are no readers currently accessing the jarFile also close it, otherwise defer the closing when the last reader
+     * will leave.
+     */
+    void close() {
+        release();
+    }
+
+    private boolean closing() {
+        return referenceCounter.get() == 0;
+    }
+
+    @FunctionalInterface
+    interface JarFileConsumer<T> {
+        T apply(JarFile jarFile, Path jarPath, String resource);
+    }
+
+    static <T> T withJarFile(AtomicReference<JarFileReference> jarFileReference, Path jarPath, String resource,
+            JarFileConsumer<T> f) {
+        JarFileReference localJarFileRef = get(jarFileReference, jarPath);
+        JarFile jarFile = localJarFileRef.acquire();
+        while (jarFile == null) {
+            JarFileReference newRef = get(jarFileReference, jarPath);
+            if (newRef != localJarFileRef) {
+                localJarFileRef = newRef;
+                jarFile = localJarFileRef.acquire();
+            }
+        }
+
+        try {
+            return f.apply(jarFile, jarPath, resource);
+        } finally {
+            localJarFileRef.release();
+        }
+    }
+
+    private static JarFileReference get(AtomicReference<JarFileReference> jarFileReference, Path jarPath) {
+        JarFileReference localJarFileRef = jarFileReference.get();
+        if (localJarFileRef != null && !localJarFileRef.closing()) {
+            return localJarFileRef;
+        }
+        JarFileReference newJarFileRef = JarFileReference.loadJarFile(jarFileReference, jarPath);
+        return jarFileReference.compareAndSet(localJarFileRef, newJarFileRef) ? newJarFileRef : jarFileReference.get();
```

Comment:
> jarFileReference.get() is not guaranteed here to return anything, it could be null. If the compare and set fail maybe another thread has succeeded to close and null out the reference. We should keep on trying until we succeed to create and set or get something not closed or not null.

I fixed this, with another loop.

> I suggest too, while creating the new JarFileReference to have fast path which lazySet the reference counter to 2 (which means created + acquire) before performing the compare and set on the atomic reference.

I tried to implement this suggestion, but if I set the counter to 2 when `JarFileReference` is created I don't see a way to distinguish this situation with the one when the reference counter is 2 and a subsequent acquire is requested, after in the `JarFileReference` lifecycle. The only way I see to introduce a fast path for the first acquire would be with a second state variable, which however is exactly the thing that we want to avoid.

Reply:
Let's talk about it tomorrow, in flesh :P